components/sync15/src/telemetry.rs

```diff
@@ -0,0 +1,645 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Manage recording sync telemetry. Assumes some external telemetry
+// library/code which manages submitting.
+
+use std::collections::HashMap;
+use std::time;
+
+use num::Num;
+
+// We use serde to serialize to json (but we never need to deserialize)
+use serde::ser::{Serialize, Serializer};
+use serde_derive::*;
+#[cfg(test)]
+use serde_json::{self, json};
+
+use crate::error::Error;
+
+// For skip_serializing_if
+// I'm surprised I can't use Num::is_zero directly, although Option::is_none works.
+// (Num's just a trait, Option is an enum, but surely there's *some* way?)
+fn skip_if_zero<T: Num + PartialEq>(v: &T) -> bool {
+    return v.is_zero();
```

Comment:
Nit: this shouldn' t have an explicit return.

Actually, while I'm here, num is an awfully heavyweight dep to pull in for this function... It's pretty slow to compile.

Another way to write something roughly equivalent for our uses might be:

```rust
fn skip_if_default<T: PartialEq + Default>(v: T) -> bool {
    v == T::default()
}
```

You could probably also do something like `v.into() == 0u64` since it seems like you only use it for usize/u32 which should be convertable to that, but this is a bit more general.